ttnn/cpp/ttnn/operations/pool/pool_utils.cpp

```diff
@@ -41,4 +47,253 @@ std::map<std::string, std::string> get_defines(Pool2DType pool_type) {
 
     return defines;
 }
-}  // namespace ttnn::operations::pool
+
+using sliding_window::ParallelConfig;
+using sliding_window::SlidingWindowConfig;
+
+std::optional<ParallelConfig> determine_valid_parallel_config(
+    const TensorMemoryLayout shard_layout,
+    uint32_t batch_size,
+    uint32_t channels,
+    uint32_t output_height,
+    uint32_t output_width,
+    const CoreCoord& compute_grid_size,
+    ShardOrientation block_shard_orientation,
+    bool enable_channels_padding,
+    bool is_shard_height_tile_multiple,
+    bool is_shard_width_tile_multiple,
+    uint32_t act_block_h_override) {
+    if (shard_layout != TensorMemoryLayout::HEIGHT_SHARDED && shard_layout != TensorMemoryLayout::BLOCK_SHARDED &&
+        shard_layout != TensorMemoryLayout::WIDTH_SHARDED) {
+        TT_THROW("Pool2d supports Height, Block or Width Sharded Layouts but got {}", shard_layout);
+    }
+    auto pconfig = conv::determine_parallel_config(
+        shard_layout,
+        batch_size,
+        channels,
+        output_height,
+        output_width,
+        channels,
+        compute_grid_size,
+        block_shard_orientation,
+        enable_channels_padding,
+        is_shard_height_tile_multiple,
+        is_shard_width_tile_multiple,
+        act_block_h_override);
+
+    // pooling can accept any height and either a tile multiple or half a tile for width.
+    if (shard_layout == TensorMemoryLayout::HEIGHT_SHARDED) {
+        uint32_t num_cores_c = 1;
+        uint32_t tile_size = channels / num_cores_c;
+        if (tile_size != 16 && tile_size % tt::constants::TILE_WIDTH != 0) {
+            return std::nullopt;
+        }
+    } else if (shard_layout == TensorMemoryLayout::BLOCK_SHARDED) {
+        auto grid_x = pconfig.grid.ranges()[0].end_coord.x - pconfig.grid.ranges()[0].start_coord.x + 1;
+        auto grid_y = pconfig.grid.ranges()[0].end_coord.y - pconfig.grid.ranges()[0].start_coord.y + 1;
+        uint32_t num_cores_c = block_shard_orientation == ShardOrientation::COL_MAJOR ? grid_y : grid_x;
+        uint32_t tile_size = channels / num_cores_c;
+        if (tile_size != 16 && tile_size % tt::constants::TILE_WIDTH != 0) {
+            return std::nullopt;
```

Comment:
There is no point to return nullopt here and than assert on the outside that this is not nullopt. Just add asserts here

Reply:
It does happen in ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp line 90